pkg/util/kubernetes/apiserver/hpa_controler.go

```diff
@@ -0,0 +1,312 @@
+	var err error
+	h := &AutoscalersController{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "autoscalers"),
+	}
+
+	h.poller = Poller{
+		gcPeriodSeconds: config.Datadog.GetInt("hpa_watcher_gc_period"),
+		refreshPeriod:   config.Datadog.GetInt("external_metrics_provider.polling_freq"),
+		batchWindow:     config.Datadog.GetInt("external_metrics_provider.batch_window"),
+	}
+
+	// Setup the client to process the HPA and metrics
+	h.hpaProc, err = hpa.NewHPAProcessor(dogCl)
+	if err != nil {
+		log.Errorf("Could not instantiate the HPA Processor: %v", err.Error())
+		return nil
+	}
+	h.clientSet = client
+	h.le = le // only trigger GC and updateExternalMetrics by the Leader.
+
+	datadogHPAConfigMap := custommetrics.GetConfigmapName()
+	h.store, err = custommetrics.NewConfigMapStore(client, common.GetResourcesNamespace(), datadogHPAConfigMap)
+	if err != nil {
+		log.Errorf("Could not instantiate the local store for the External Metrics %v", err)
+		return nil
+	}
+
+	autoscalingInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    h.addAutoscaler,
+			UpdateFunc: h.updateAutoscaler,
+			DeleteFunc: h.deleteAutoscaler,
+		},
+	)
+	h.autoscalersLister = autoscalingInformer.Lister()
+	h.autoscalersListerSynced = autoscalingInformer.Informer().HasSynced
+	return h
+}
+
+func (h *AutoscalersController) Run(stopCh <-chan struct{}) {
+	defer h.queue.ShutDown()
+
+	log.Infof("Starting HPA Controller ... ")
+	defer log.Infof("Stopping HPA Controller")
+
+	if !cache.WaitForCacheSync(stopCh, h.autoscalersListerSynced) {
+		return
+	}
+
+	h.processingLoop()
+
+	go wait.Until(h.worker, time.Second, stopCh)
+	<-stopCh
+}
+
+// processingLoop is a second go routing that schedules the garbage collection and the refreshing of external metrics
+// in the GlobalStore.
+func (c *AutoscalersController) processingLoop() {
+	tickerHPARefreshProcess := time.NewTicker(time.Duration(c.poller.refreshPeriod) * time.Second)
+	gcPeriodSeconds := time.NewTicker(time.Duration(c.poller.gcPeriodSeconds) * time.Second)
+	batchFreq := time.NewTicker(time.Duration(c.poller.batchWindow) * time.Second)
+
+	go func() {
+		for {
+			select {
+			case <-tickerHPARefreshProcess.C:
+				if !c.le.IsLeader() {
+					c.toStore.m.Lock()
+					c.toStore.data = nil
+					c.toStore.m.Unlock()
+					continue
+				}
+				// Updating the metrics against Datadog should not affect the HPA pipeline.
+				// If metrics are temporarily unavailable for too long, they will become `Valid=false` and won't be evaluated.
+				c.updateExternalMetrics()
+			case <-gcPeriodSeconds.C:
+				if !c.le.IsLeader() {
+					continue
+				}
+				c.gc()
+			case <-batchFreq.C:
+				c.toStore.m.Lock()
+				data := c.toStore.data
+				c.toStore.m.Unlock()
+				if !c.le.IsLeader() || len(data) == 0 {
```

Comment:
What having `pushToGlobalStore()` take no parameters. The `pushToGlobalStore()` can look int the `c.toStore` itself and decide what to do based on whether `c.le.IsLeader()`. 

Regardless of whether we are the leader, this function should clear the `c.toStore` but if we are the leader we save the data in the configmap.